R/downloader.R

```diff
@@ -37,32 +37,49 @@ download <- function(url, ...) {
   # First, check protocol. If http or https, check platform:
   if (grepl('^https?://', url)) {
 
-    # If Windows, call setInternet2, then use download.file with defaults.
+    # Check whether we are running R 3.2 and whether we have libcurl
+    isR32 <- getRversion() >= "3.2"
+
+    # Windows
     if (.Platform$OS.type == "windows") {
 
-      # If we directly use setInternet2, R CMD CHECK gives a Note on Mac/Linux
-      seti2 <- `::`(utils, 'setInternet2')
+      if (isR32) {
+        method <- "wininet"
+      } else {
 
-      # Store initial settings, and restore on exit
-      internet2_start <- seti2(NA)
-      on.exit(suppressWarnings(seti2(internet2_start)))
+        # If we directly use setInternet2, R CMD CHECK gives a Note on Mac/Linux
+        seti2 <- `::`(utils, 'setInternet2')
 
-      # Needed for https. Will get warning if setInternet2(FALSE) already run
-      # and internet routines are used. But the warnings don't seem to matter.
-      suppressWarnings(seti2(TRUE))
+        # Check whether we are already using internet2 for internal
+        internet2_start <- seti2(NA)
+
+        # If not then temporarily set it
+        if (!internet2_start) {
+          # Store initial settings, and restore on exit
+          on.exit(suppressWarnings(seti2(internet2_start)))
+
+          # Needed for https. Will get warning if setInternet2(FALSE) already run
+          # and internet routines are used. But the warnings don't seem to matter.
+          suppressWarnings(seti2(TRUE))
+        }
+
+        method <- "internal"
+      }
 
       # download.file will complain about file size with something like:
       #       Warning message:
       #         In download.file(url, ...) : downloaded length 19457 != reported length 200
       # because apparently it compares the length with the status code returned (?)
       # so we supress that
-      suppressWarnings(download.file(url, ...))
+      suppressWarnings(download.file(url, method = method, ...))
 
     } else {
-      # If non-Windows, check for curl/wget/lynx, then call download.file with
+      # If non-Windows, check for libcurl/curl/wget/lynx, then call download.file with
       # appropriate method.
 
-      if (nzchar(Sys.which("wget")[1])) {
+      if (isR32 && capabilities("libcurl")) {
```

Comment:
Yes, that's correct.

On Tue, Jul 7, 2015 at 2:30 PM, Kevin Ushey <notifications@github.com>
wrote:

> In R/downloader.R
> <https://github.com/rstudio/packrat/pull/230#discussion_r34073563>:
>
> >        # appropriate method.
> >
> > -      if (nzchar(Sys.which("wget")[1])) {
> > +      if (isR32 && capabilities("libcurl")) {
>
> I guess since the check only occurs after an isR32 call it's probably
> fine?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/packrat/pull/230/files#r34073563>.
>